// backendless/es/messaging/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _dec, _dec2, _dec3, _dec4, _dec5, _dec6, _dec7, _dec8, _desc, _value, _obj, _init, _init2, _init3, _init4, _init5, _init6, _init7, _init8;

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _utils = require('../utils');

var _utils2 = _interopRequireDefault(_utils);

var _decorators = require('../decorators/index');

var _channel = require('./channel/index');

var _channel2 = _interopRequireDefault(_channel);

var _helpers = require('./helpers/index');

var _methods = require('./methods/index');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _applyDecoratedDescriptor(target, property, decorators, descriptor, context) {
  var desc = {};
  Object['ke' + 'ys'](descriptor).forEach(function (key) {
    desc[key] = descriptor[key];
  });
  desc.enumerable = !!desc.enumerable;
  desc.configurable = !!desc.configurable;

  if ('value' in desc || desc.initializer) {
    desc.writable = true;
  }

  desc = decorators.slice().reverse().reduce(function (desc, decorator) {
    return decorator(target, property, desc) || desc;
  }, desc);

  if (context && desc.initializer !== void 0) {
    desc.value = desc.initializer ? desc.initializer.call(context) : void 0;
    desc.initializer = undefined;
  }

  if (desc.initializer === void 0) {
    Object['define' + 'Property'](target, property, desc);
    desc = null;
  }

  return desc;
}

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

var Messaging = function () {
  function Messaging(app) {
    _classCallCheck(this, Messaging);

    this.app = app;

    this.Bodyparts = _helpers.Bodyparts;
    this.PublishOptions = _helpers.PublishOptions;
    this.DeliveryOptions = _helpers.DeliveryOptions;
    this.PublishOptionsHeaders = _helpers.PublishOptionsHeaders;
    this.EmailEnvelope = _helpers.EmailEnvelope;

    /** @deprecated */
    this.SubscriptionOptions = _helpers.SubscriptionOptions;
  }

  _createClass(Messaging, [{
    key: 'subscribe',
    value: function subscribe(channelName) {
      if (!channelName || typeof channelName !== 'string') {
        throw new Error('"channelName" must be non empty string');
      }

      if (channelName.indexOf('/') >= 0) {
        throw new Error('"channelName" can not contains slash chars');
      }

      return new _channel2.default({ name: channelName.trim() }, this.app);
    }
  }]);

  return Messaging;
}();

_extends(Messaging.prototype, (_dec = (0, _decorators.deprecated)('Backendless.Messaging', 'Backendless.Messaging.publish'), _dec2 = (0, _decorators.deprecated)('Backendless.Messaging', 'Backendless.Messaging.sendEmail'), _dec3 = (0, _decorators.deprecated)('Backendless.Messaging', 'Backendless.Messaging.sendEmailFromTemplate'), _dec4 = (0, _decorators.deprecated)('Backendless.Messaging', 'Backendless.Messaging.cancel'), _dec5 = (0, _decorators.deprecated)('Backendless.Messaging', 'Backendless.Messaging.registerDevice'), _dec6 = (0, _decorators.deprecated)('Backendless.Messaging', 'Backendless.Messaging.getRegistrations'), _dec7 = (0, _decorators.deprecated)('Backendless.Messaging', 'Backendless.Messaging.unregisterDevice'), _dec8 = (0, _decorators.deprecated)('Backendless.Messaging', 'Backendless.Messaging.getMessageStatus'), (_obj = {
  publishSync: _utils2.default.synchronized(_methods.publish),
  publish: _utils2.default.promisified(_methods.publish),

  pushWithTemplate: _utils2.default.promisified(_methods.pushWithTemplate),

  sendEmailSync: _utils2.default.synchronized(_methods.sendEmail),
  sendEmail: _utils2.default.promisified(_methods.sendEmail),

  sendEmailFromTemplateSync: _utils2.default.synchronized(_methods.sendEmailFromTemplate),
  sendEmailFromTemplate: _utils2.default.promisified(_methods.sendEmailFromTemplate),

  cancelSync: _utils2.default.synchronized(_methods.cancel),
  cancel: _utils2.default.promisified(_methods.cancel),

  registerDeviceSync: _utils2.default.synchronized(_methods.registerDevice),
  registerDevice: _utils2.default.promisified(_methods.registerDevice),

  getRegistrationsSync: _utils2.default.synchronized(_methods.getRegistrations),
  getRegistrations: _utils2.default.promisified(_methods.getRegistrations),

  unregisterDeviceSync: _utils2.default.synchronized(_methods.unregisterDevice),
  unregisterDevice: _utils2.default.promisified(_methods.unregisterDevice),

  getMessageStatusSync: _utils2.default.synchronized(_methods.getMessageStatus),
  getMessageStatus: _utils2.default.promisified(_methods.getMessageStatus),

  getPushTemplates: _utils2.default.promisified(_methods.getPushTemplates)

}, (_applyDecoratedDescriptor(_obj, 'publishSync', [_dec], (_init = Object.getOwnPropertyDescriptor(_obj, 'publishSync'), _init = _init ? _init.value : undefined, {
  enumerable: true,
  configurable: true,
  writable: true,
  initializer: function initializer() {
    return _init;
  }
}), _obj), _applyDecoratedDescriptor(_obj, 'sendEmailSync', [_dec2], (_init2 = Object.getOwnPropertyDescriptor(_obj, 'sendEmailSync'), _init2 = _init2 ? _init2.value : undefined, {
  enumerable: true,
  configurable: true,
  writable: true,
  initializer: function initializer() {
    return _init2;
  }
}), _obj), _applyDecoratedDescriptor(_obj, 'sendEmailFromTemplateSync', [_dec3], (_init3 = Object.getOwnPropertyDescriptor(_obj, 'sendEmailFromTemplateSync'), _init3 = _init3 ? _init3.value : undefined, {
  enumerable: true,
  configurable: true,
  writable: true,
  initializer: function initializer() {
    return _init3;
  }
}), _obj), _applyDecoratedDescriptor(_obj, 'cancelSync', [_dec4], (_init4 = Object.getOwnPropertyDescriptor(_obj, 'cancelSync'), _init4 = _init4 ? _init4.value : undefined, {
  enumerable: true,
  configurable: true,
  writable: true,
  initializer: function initializer() {
    return _init4;
  }
}), _obj), _applyDecoratedDescriptor(_obj, 'registerDeviceSync', [_dec5], (_init5 = Object.getOwnPropertyDescriptor(_obj, 'registerDeviceSync'), _init5 = _init5 ? _init5.value : undefined, {
  enumerable: true,
  configurable: true,
  writable: true,
  initializer: function initializer() {
    return _init5;
  }
}), _obj), _applyDecoratedDescriptor(_obj, 'getRegistrationsSync', [_dec6], (_init6 = Object.getOwnPropertyDescriptor(_obj, 'getRegistrationsSync'), _init6 = _init6 ? _init6.value : undefined, {
  enumerable: true,
  configurable: true,
  writable: true,
  initializer: function initializer() {
    return _init6;
  }
}), _obj), _applyDecoratedDescriptor(_obj, 'unregisterDeviceSync', [_dec7], (_init7 = Object.getOwnPropertyDescriptor(_obj, 'unregisterDeviceSync'), _init7 = _init7 ? _init7.value : undefined, {
  enumerable: true,
  configurable: true,
  writable: true,
  initializer: function initializer() {
    return _init7;
  }
}), _obj), _applyDecoratedDescriptor(_obj, 'getMessageStatusSync', [_dec8], (_init8 = Object.getOwnPropertyDescriptor(_obj, 'getMessageStatusSync'), _init8 = _init8 ? _init8.value : undefined, {
  enumerable: true,
  configurable: true,
  writable: true,
  initializer: function initializer() {
    return _init8;
  }
}), _obj)), _obj)));

exports.default = Messaging;
