// backendless/es/users/update.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.update = update;

var _utils = require('../utils');

var _utils2 = _interopRequireDefault(_utils);

var _async = require('../request/async');

var _async2 = _interopRequireDefault(_async);

var _user = require('./user');

var _user2 = _interopRequireDefault(_user);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function update(user /** async */) {
  var responder = _utils2.default.extractResponder(arguments);

  var isAsync = !!responder;

  var result = this.app.request.put({
    url: this.app.urls.userObject(user.objectId),
    isAsync: isAsync,
    asyncHandler: responder && wrapAsync(responder),
    data: user
  });

  return isAsync ? result : parseResponse(result);
}

function parseResponse(data) {
  return _utils2.default.deepExtend(new _user2.default(), data);
}

function wrapAsync(asyncHandler) {
  //TODO: will be removed after remove all the "{methodName}Sync" methods

  var onSuccess = function onSuccess(data) {
    return asyncHandler.success(parseResponse(data));
  };
  var onError = function onError(error) {
    return asyncHandler.fault(error);
  };

  return new _async2.default(onSuccess, onError);
}
