// backendless/lib/messaging/methods/register-device.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.registerDevice = registerDevice;

var _utils = require('../../utils');

var _utils2 = _interopRequireDefault(_utils);

var _async = require('../../request/async');

var _async2 = _interopRequireDefault(_async);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function registerDevice(deviceToken, channels, expiration, asyncHandler) {
  var device = this.app.device;

  asyncHandler = _utils2.default.extractResponder(arguments);

  var data = {
    deviceToken: deviceToken,
    deviceId: device.uuid,
    os: device.platform,
    osVersion: device.version
  };

  if (_utils2.default.isArray(channels)) {
    data.channels = channels;
  }

  if (_utils2.default.isNumber(expiration) || expiration instanceof Date) {
    data.expiration = expiration instanceof Date ? expiration.getTime() / 1000 : expiration;
  }

  return this.app.request.post({
    url: this.app.urls.messagingRegistrations(),
    data: data,
    isAsync: !!asyncHandler,
    asyncHandler: asyncHandler
  });
}
