// backendless/lib/users/register.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.register = register;

var _utils = require('../utils');

var _utils2 = _interopRequireDefault(_utils);

var _async = require('../request/async');

var _async2 = _interopRequireDefault(_async);

var _user = require('./user');

var _user2 = _interopRequireDefault(_user);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function register(user /** async */) {
  var responder = _utils2.default.extractResponder(arguments);

  var isAsync = !!responder;

  var result = this.app.request.post({
    url: this.app.urls.userRegister(),
    isAsync: isAsync,
    asyncHandler: responder && wrapAsync(responder),
    data: enrichWithLocaleInfo(user)
  });

  return isAsync ? result : parseResponse(result);
}

function enrichWithLocaleInfo(user) {
  if (!user.blUserLocale) {
    var clientUserLocale = getClientUserLocale();

    if (clientUserLocale) {
      user.blUserLocale = clientUserLocale;
    }
  }

  return user;
}

function getClientUserLocale() {
  if (typeof navigator === 'undefined') {
    return;
  }

  var language = '';

  if (navigator.languages && navigator.languages.length) {
    language = navigator.languages[0];
  } else {
    language = navigator.userLanguage || navigator.language || navigator.browserLanguage || navigator.systemLanguage || '';
  }

  return language.slice(0, 2).toLowerCase();
}

function parseResponse(data) {
  return _utils2.default.deepExtend(new _user2.default(), data);
}

function wrapAsync(asyncHandler) {
  var onSuccess = function onSuccess(data) {
    return asyncHandler.success(parseResponse(data));
  };
  var onError = function onError(error) {
    return asyncHandler.fault(error);
  };

  return new _async2.default(onSuccess, onError);
}
