// backendless/es/files/upload.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.upload = upload;

var _async = require('../request/async');

var _async2 = _interopRequireDefault(_async);

var _send = require('./send');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var getFileName = function getFileName(file) {
  if (file.name) {
    return file.name;
  }

  if (file.path) {
    var path = file.path.split('/');
    return path[path.length - 1]; //last item of the file path
  }
};

/**
 * @param {File} file
 * @param {String} filePath
 * @param {Boolean} overwrite
 * @param {Async} asyncHandler
 * @returns {Promise.<String>}
 */
function upload(file, filePath, overwrite, asyncHandler) {
  var fileName = getFileName(file);

  if (!fileName) {
    throw new Error('Wrong type of the file source object. Can not get file name');
  }

  if (overwrite instanceof _async2.default) {
    asyncHandler = overwrite;
    overwrite = undefined;
  }

  return _send.sendFile.call(this, {
    overwrite: overwrite,
    path: filePath,
    fileName: fileName,
    file: file,
    asyncHandler: asyncHandler
  });
}
