// backendless/es/files/persmissions/persmission.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = undefined;

var _dec, _dec2, _dec3, _dec4, _dec5, _dec6, _dec7, _dec8, _dec9, _dec10, _dec11, _dec12, _desc, _value, _class, _descriptor, _descriptor2, _descriptor3, _descriptor4, _descriptor5, _descriptor6, _descriptor7, _descriptor8, _descriptor9, _descriptor10, _descriptor11, _descriptor12;

var _utils = require('../../utils');

var _utils2 = _interopRequireDefault(_utils);

var _decorators = require('../../decorators/index');

var _sendRequest = require('./send-request');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _initDefineProp(target, property, descriptor, context) {
  if (!descriptor) return;
  Object.defineProperty(target, property, {
    enumerable: descriptor.enumerable,
    configurable: descriptor.configurable,
    writable: descriptor.writable,
    value: descriptor.initializer ? descriptor.initializer.call(context) : void 0
  });
}

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _applyDecoratedDescriptor(target, property, decorators, descriptor, context) {
  var desc = {};
  Object['ke' + 'ys'](descriptor).forEach(function (key) {
    desc[key] = descriptor[key];
  });
  desc.enumerable = !!desc.enumerable;
  desc.configurable = !!desc.configurable;

  if ('value' in desc || desc.initializer) {
    desc.writable = true;
  }

  desc = decorators.slice().reverse().reduce(function (desc, decorator) {
    return decorator(target, property, desc) || desc;
  }, desc);

  if (context && desc.initializer !== void 0) {
    desc.value = desc.initializer ? desc.initializer.call(context) : void 0;
    desc.initializer = undefined;
  }

  if (desc.initializer === void 0) {
    Object['define' + 'Property'](target, property, desc);
    desc = null;
  }

  return desc;
}

function _initializerWarningHelper(descriptor, context) {
  throw new Error('Decorating class property failed. Please ensure that transform-class-properties is enabled.');
}

var PermissionTypes = {
  GRANT: 'GRANT',
  DENY: 'DENY'
};

function grantForUser(userId, url, asyncHandler) {
  return _sendRequest.sendRequest.call(this, PermissionTypes.GRANT, url, { userId: userId }, asyncHandler);
}

function denyForUser(userId, url, asyncHandler) {
  return _sendRequest.sendRequest.call(this, PermissionTypes.DENY, url, { userId: userId }, asyncHandler);
}

function grantForRole(roleName, url, asyncHandler) {
  return _sendRequest.sendRequest.call(this, PermissionTypes.GRANT, url, { roleName: roleName }, asyncHandler);
}

function denyForRole(roleName, url, asyncHandler) {
  return _sendRequest.sendRequest.call(this, PermissionTypes.DENY, url, { roleName: roleName }, asyncHandler);
}

function grantForAllUsers(url, asyncHandler) {
  return _sendRequest.sendRequest.call(this, PermissionTypes.GRANT, url, { userId: '*' }, asyncHandler);
}

function denyForAllUsers(url, asyncHandler) {
  return _sendRequest.sendRequest.call(this, PermissionTypes.DENY, url, { userId: '*' }, asyncHandler);
}

function grantForAllRoles(url, asyncHandler) {
  return _sendRequest.sendRequest.call(this, PermissionTypes.GRANT, url, { roleName: '*' }, asyncHandler);
}

function denyForAllRoles(url, asyncHandler) {
  return _sendRequest.sendRequest.call(this, PermissionTypes.DENY, url, { roleName: '*' }, asyncHandler);
}

//TODO: will be removed when remove sync methods
var namespaceLabel = 'Backendless.Files.Permissions.{READ|DELETE|WRITE}';

var FilePermission = (_dec = (0, _decorators.deprecated)(namespaceLabel, namespaceLabel + '.grantForUser'), _dec2 = (0, _decorators.deprecated)(namespaceLabel, namespaceLabel + '.grantForUser'), _dec3 = (0, _decorators.deprecated)(namespaceLabel, namespaceLabel + '.denyForUser'), _dec4 = (0, _decorators.deprecated)(namespaceLabel, namespaceLabel + '.denyForUser'), _dec5 = (0, _decorators.deprecated)(namespaceLabel, namespaceLabel + '.grantForRole'), _dec6 = (0, _decorators.deprecated)(namespaceLabel, namespaceLabel + '.grantForRole'), _dec7 = (0, _decorators.deprecated)(namespaceLabel, namespaceLabel + '.denyForRole'), _dec8 = (0, _decorators.deprecated)(namespaceLabel, namespaceLabel + '.denyForRole'), _dec9 = (0, _decorators.deprecated)(namespaceLabel, namespaceLabel + '.grantForAllUsers'), _dec10 = (0, _decorators.deprecated)(namespaceLabel, namespaceLabel + '.grantForAllUsers'), _dec11 = (0, _decorators.deprecated)(namespaceLabel, namespaceLabel + '.denyForAllUsers'), _dec12 = (0, _decorators.deprecated)(namespaceLabel, namespaceLabel + '.denyForAllUsers'), (_class = function FilePermission(permission, app) {
  _classCallCheck(this, FilePermission);

  _initDefineProp(this, 'grantUserSync', _descriptor, this);

  _initDefineProp(this, 'grantUser', _descriptor2, this);

  _initDefineProp(this, 'denyUserSync', _descriptor3, this);

  _initDefineProp(this, 'denyUser', _descriptor4, this);

  _initDefineProp(this, 'grantRoleSync', _descriptor5, this);

  _initDefineProp(this, 'grantRole', _descriptor6, this);

  _initDefineProp(this, 'denyRoleSync', _descriptor7, this);

  _initDefineProp(this, 'denyRole', _descriptor8, this);

  _initDefineProp(this, 'grantSync', _descriptor9, this);

  _initDefineProp(this, 'grant', _descriptor10, this);

  _initDefineProp(this, 'denySync', _descriptor11, this);

  _initDefineProp(this, 'deny', _descriptor12, this);

  this.grantForUser = _utils2.default.promisified(grantForUser);
  this.denyForUser = _utils2.default.promisified(denyForUser);
  this.grantForRole = _utils2.default.promisified(grantForRole);
  this.denyForRole = _utils2.default.promisified(denyForRole);
  this.grantForAllUsers = _utils2.default.promisified(grantForAllUsers);
  this.denyForAllUsers = _utils2.default.promisified(denyForAllUsers);
  this.grantForAllRoles = _utils2.default.promisified(grantForAllRoles);
  this.denyForAllRoles = _utils2.default.promisified(denyForAllRoles);

  this.permission = permission;
  this.app = app;
}, (_descriptor = _applyDecoratedDescriptor(_class.prototype, 'grantUserSync', [_dec], {
  enumerable: true,
  initializer: function initializer() {
    return _utils2.default.synchronized(grantForUser);
  }
}), _descriptor2 = _applyDecoratedDescriptor(_class.prototype, 'grantUser', [_dec2], {
  enumerable: true,
  initializer: function initializer() {
    return _utils2.default.promisified(grantForUser);
  }
}), _descriptor3 = _applyDecoratedDescriptor(_class.prototype, 'denyUserSync', [_dec3], {
  enumerable: true,
  initializer: function initializer() {
    return _utils2.default.synchronized(denyForUser);
  }
}), _descriptor4 = _applyDecoratedDescriptor(_class.prototype, 'denyUser', [_dec4], {
  enumerable: true,
  initializer: function initializer() {
    return _utils2.default.promisified(denyForUser);
  }
}), _descriptor5 = _applyDecoratedDescriptor(_class.prototype, 'grantRoleSync', [_dec5], {
  enumerable: true,
  initializer: function initializer() {
    return _utils2.default.synchronized(grantForRole);
  }
}), _descriptor6 = _applyDecoratedDescriptor(_class.prototype, 'grantRole', [_dec6], {
  enumerable: true,
  initializer: function initializer() {
    return _utils2.default.promisified(grantForRole);
  }
}), _descriptor7 = _applyDecoratedDescriptor(_class.prototype, 'denyRoleSync', [_dec7], {
  enumerable: true,
  initializer: function initializer() {
    return _utils2.default.synchronized(denyForRole);
  }
}), _descriptor8 = _applyDecoratedDescriptor(_class.prototype, 'denyRole', [_dec8], {
  enumerable: true,
  initializer: function initializer() {
    return _utils2.default.promisified(denyForRole);
  }
}), _descriptor9 = _applyDecoratedDescriptor(_class.prototype, 'grantSync', [_dec9], {
  enumerable: true,
  initializer: function initializer() {
    return _utils2.default.synchronized(grantForAllUsers);
  }
}), _descriptor10 = _applyDecoratedDescriptor(_class.prototype, 'grant', [_dec10], {
  enumerable: true,
  initializer: function initializer() {
    return _utils2.default.promisified(grantForAllUsers);
  }
}), _descriptor11 = _applyDecoratedDescriptor(_class.prototype, 'denySync', [_dec11], {
  enumerable: true,
  initializer: function initializer() {
    return _utils2.default.synchronized(denyForAllUsers);
  }
}), _descriptor12 = _applyDecoratedDescriptor(_class.prototype, 'deny', [_dec12], {
  enumerable: true,
  initializer: function initializer() {
    return _utils2.default.promisified(denyForAllUsers);
  }
})), _class));
exports.default = FilePermission;
