// backendless/es/users/social/request.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.sendSocialLoginRequest = sendSocialLoginRequest;

var _async = require('../../request/async');

var _async2 = _interopRequireDefault(_async);

var _utils = require('../utils');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function sendSocialLoginRequest(accessToken, socialType, fieldsMapping, stayLoggedIn, asyncHandler) {
  if (!accessToken) {
    return asyncHandler.fault('"accessToken" is missing.');
  }

  var context = this;

  var interimCallback = new _async2.default(function (r) {
    context.setLocalCurrentUser(_utils.parseResponse.call(context, r));
    context.app.LocalCache.set('stayLoggedIn', !!stayLoggedIn);

    asyncHandler.success(_utils.getUserFromResponse.call(context, context.getLocalCurrentUser()));
  }, function (e) {
    asyncHandler.fault(e);
  });

  this.app.request.post({
    url: this.app.urls.userSocialLogin(socialType),
    isAsync: true,
    asyncHandler: interimCallback,
    data: {
      accessToken: accessToken,
      fieldsMapping: fieldsMapping
    }
  });
}
